noxfile.py

```diff
@@ -195,27 +241,55 @@ def build_release(session):
 
     session.log("# Ensure no files in dist/")
     if release.have_files_in_folder("dist"):
-        session.error("There are files in dist/. Remove them and try again")
+        session.error(
+            "There are files in dist/. Remove them and try again. "
+            "You can use `git clean -fxdi -- dist` command to do this"
+        )
 
     session.log("# Install dependencies")
     session.install("setuptools", "wheel", "twine")
 
-    session.log("# Checkout the tag")
-    session.run("git", "checkout", version, external=True, silent=True)
+    with mk_tmp_git_checkout(session, version) as build_dir_path:
+        with workdir(session, build_dir_path):
+            build_dists(session)
+
+        tmp_dist_dir = build_dir_path / 'dist'
+        session.log(f"# Copying dists from tmp {tmp_dist_dir}")
+        shutil.copytree(tmp_dist_dir, 'dist')
 
+
+def build_dists(session):
     session.log("# Cleanup build/ before building the wheel")
     if release.have_files_in_folder("build"):
         shutil.rmtree("build")
 
+    session.log(
+        "# Check if there's any Git-untracked files before building the wheel",
+    )
+    has_git_untracked_files = any(
+        bool(l) for l in
+        # session.run doesn't seem to return any output
+        subprocess.check_output(
+            (
+                "git", "ls-files", "--ignored", "--exclude-standard",
+                "--others", "--", ".",
+            ),
+            text=True,
+        ).split('\n')
+        if not l.startswith('.nox/build-release/')  # exclude nox env file
+    )
```

Comment:
(Nit) I would refactor the `any(...)` part into a separate generator function; this comprehension block is a bit difficult to read.

Reply:
@uranusjr right, makes sense. I'll do some refactoring. This was just an attempt to demo the top-level approach before polishing things.